### ISSUE TITLE
Slow data channels?

### ISSUE THREAD

**Issue by cw-rterm** [shinyoshiaki/werift-webrtc#310]
First of all, great job on this library. As a long time WebRTC and Javascript/Typescript fan I have been searching for something like this for a long time. A library with no external dependencies, amazing!

One thing I noticed is that the data channels are quite slow, at least compared to the `wrtc` node library. Am I doing something wrong or is this a known issue?

I ran a quick 30 second test uploading an 800mb file, it did not upload the whole file in either case (although I am on fiber here). The results are as follows
- wrtc: 70MB in 30 seconds
- werift: 3.4MB in 30 seconds

I wouldn't consider the `wrtc` fast, although it does perform quite a bit better than `werift` in my quick test here. 

The setup:
- Digital Ocean Server and MacOS client (both on 1GB/s internet connections)
- I have an "agent" on the server and a cli on my Mac, both running `wrtc` or `werift` on each side. 
- The file is on my Mac being sent to the Digital Ocean server (SFO2) and I am in Seattle, WA (not too far away)

The code:
- To upload the file, I am sending chunks of various sizes. For the `wrtc` library I have to be quite careful because it crashes if the buffer fills up.
- `wrtc` is the fastest when i send about 5000 byte chunks (which I have always found strange) to the datachannel
- `werift` seems to like 65000 byte chunks (although 1mb chunks performed about the same)
- I stop submitting to the buffer if it is over 16438932 bytes, and start again when the datachannel buffer drops below that number.
- on the server side, I am monitoring the events right off the datachannel, to rule out any issues on that end. The server writes to the disk using Node stream, and based on the logs it takes a few milliseconds to write then waits again for the event to fire.

Thanks again for all the work on this library. Let me know if you do have any suggestions or if I can clarify more on the setup/code.

**Comment by tsightler** [comment]
I don't know the answer to your question, but my project uses werift via a dependency on another project and, since the switch to werift, I've struggled to determine if it is fast enough.  The streams I'm processing are about 3Mbps, and it seems like it can barely keep up.  Still, I think that's better than the results you posted.

Basically, I've found werift to be incredibly reliable and overall compatible, but performance, that's an entirely different story.  I've tried to spend some time with the code and, honestly, I didn't see a lot of places where it could be improved, but I'm not a very experienced Javascript developer and I'm in way over my head, however, but the overall packet path is pretty straightforward.  When I did some performance traces the biggest overhead seemed to be in the cost of decrypting SRTP using CTR encryption because it requires a call to createDecipheriv() for every packet, which just seems costly no matter what.

Summary, it's an amazing project, I've learned a ton from it, but I'm just not sure if Typescript/Javascript is fast enough to really be a streaming engine for hard core use.  That being said, between my project and other projects built using the dependency, there are literally 10,000+ users, so I guess that alone says it's good enough for that case, even if just barely.  On the other hand, if I needed a higher bitrate, I don't think it would work, except perhaps on higher end hardware, although, even there, I struggle to get much performance.  On Linux, increasing the UDP buffer at the OS level helped a good bit, but it's not always possible (for example, when running in Docker or without privileges).  I really wish NodeJS had native WebRTC functionality.

**Comment by koush** [comment]
For what it's worth, I'm using werift to handle multiple 4k video streams (16Mbps) with no issue (which uses rtp). However, the datachannels do seem slow. I believe this is due to either the sctp implementation or sctp itself. I haven't looked into it as much yet.

**Comment by tsightler** [comment]
Yep, I'm pretty familiar with Scrypted, have also run it myself.  I think we might just have different definitions of "runs great", I would consider it as "runs sufficiently", probably as good as could be done with JS.  I'm probably just comparing it with the Javascript RTP implementations I used prior to WebRTC, where the same streams used 50-65% less CPU for the same bitrate.  The traces I took comparing the two were pretty clear that the difference was almost exclusively due to additional decryption overhead.  But compared to something native, werift isn't a great performer, that's not really a fault of werift itself though.

**Comment by koush** [comment]
The decrypt/encrypt functions all uses openssl/native code under the hood, so I'm a little skeptical that's actually the case?

**Comment by tsightler** [comment]
Fair enough, I didn't say that was for sure the issue, it was just the only obvious differences in the traces compared to the previous RTP code I was using, and it was very clear that a lot more time was spent in the decrypt function, and that seemed to explain the overall significantly higher CPU usage (to be fair, it was completely different cipher that was overall less complex in general).

I understand that native code is still used of course (in my daily job I work with C++ code that encrypts data at 100's of MB/s), but any time you have to initiate a new decipher object for every object being decrypted vs using an already existing decipher object, it's quite noticeable (some synthetic benchmarks show 100x faster for the latter), so, even though it's only happening perhaps a few dozen times/second, it's still measureable.  I believe this is clearly the cause of the extra CPU overhead, but, perhaps not that impactful to the overall performance of the pipeline.  But I admittedly didn't keep digging as I didn't really have any control over it and have just tried to optimize my side of the code as much as possible.

**Comment by koush** [comment]
@tsightler incidentally, are you primarily sending or receiving data with werift?

**Comment by tsightler** [comment]
Receiving.  Basically, my use is via ring-client-api, receiving streams from Ring and piping them out via RTSP.  Overall, werift works fairly well with this use case, but overall CPU use is pretty high compared to the bitrate of the streams being processed, and I don't find performance amazing.

**Comment by koush** [comment]
Ah gotcha. I implemented the non-BUNDLE transport support in werift for `ring-client-api`, and have an RTSP gateway for Ring in scrypted as well. My primary werift use case is sending streams with Scrypted, so I never did a performance pass on receiving webrtc streams. A quick glance at `decryptRtp` on CTR encrypted sessions (which ring uses) and it seems there could be some optimizations could be made:

```ts
  decryptRtp(cipherText: Buffer, rolloverCounter: number): [Buffer, RtpHeader] {
    const header = RtpHeader.deSerialize(cipherText);

    let dst = Buffer.from([]);
    dst = growBufferSize(dst, cipherText.length - this.authTagLength);

    cipherText = cipherText.subarray(0, cipherText.length - this.authTagLength);

    cipherText.subarray(0, header.payloadOffset).copy(dst);

    const counter = this.generateCounter(
      header.sequenceNumber,
      rolloverCounter,
      header.ssrc,
      this.srtpSessionSalt
    );
    const cipher = createDecipheriv(
      "aes-128-ctr",
      this.srtpSessionKey,
      counter
    );
    const payload = cipherText.slice(header.payloadOffset);
    const buf = cipher.update(payload);
    buf.copy(dst, header.payloadOffset);

    return [dst, header];
  }
```

`ring-client-api` also uses `rxjs` to deliver the packets which is likely adding a few layers onto the call stack. I don't use that for video processing for that reason. https://github.com/dgreif/ring/blob/7fdeafdd853b55be59b7b729a3fb79c33485b58f/packages/ring-client-api/streaming/streaming-session.ts#L154

**Comment by koush** [comment]
For what it's worth, I found ring's webrtc implementation underwhelming/unreliable compared to SIP. So I added that back in in a forked `ring-client-api`. I suspect Amazon is running a SIP to webrtc gateway on their backend. I still use Webrtc for ring edge devices, as that is the only option (and provides a local transport).

**Comment by tsightler** [comment]
I've been recently playing with using werift directly exactly for that reason (rxjs) but, so far, I haven't found that to be a huge part of the overall performance issue to this point.  I've actually considered just re-implementing SIP in my side of the code as well, but instead I actually split the werift streams into dedicated worker threads so that they can at least be scheduled across all cores independent of the main thread.  That has managed to get similar-ish overall performance, just with much higher overall CPU usage.

However, I did some hackish test with golang and Pion, and the performance was so much better, and lighter on CPU and RAM, it makes me think I just need to spend some time to try to learn that well enough to write something usable.  But, as this is just a side hacking project, I'm not sure when or if I'll ever get that much time.

**Comment by koush** [comment]
Taking a look at the js profiler, the werift nack loop is unnecessarily running 200 times per second, even when there is no packet loss. The top hit:

<img width="447" alt="image" src="https://user-images.githubusercontent.com/73924/212201746-eff1545b-0516-4429-b0f7-2d5158458d2e.png">


```ts
  private start() {
    if (this.nackLoop) {
      return;
    }
    this.nackLoop = setInterval(async () => {
      try {
        await this.sendNack();
      } catch (error) {
        log("failed to send nack", error);
      }
    }, 5);
  }
```

**Comment by koush** [comment]
> but instead I actually split the werift streams into dedicated worker threads so that they can at least be scheduled across all cores independent of the main thread 

Using separate worker threads is what I do in Scrypted as well, it schedules nicer with multiple cores and multiple streams.

**Comment by koush** [comment]
Fixed the nack handler to use a resetting timeout:

<img width="421" alt="image" src="https://user-images.githubusercontent.com/73924/212202311-82c18096-8488-4059-9cfa-b4c772f79687.png">

**Comment by koush** [comment]
Fixed the nack handler to use a resetting timeout when a nack actually needs to happen.

<img width="421" alt="image" src="https://user-images.githubusercontent.com/73924/212202441-e8f4645d-c7b2-4fc3-8348-708984eb7e2b.png">

**Comment by cw-rterm** [comment]
I will be honest, most of this is over my head, but I do spend my days as a Typescript/Javascript dev, so will follow along where I can. I am avid fan of everything WebRTC though!

The one optimization I was able to find (by poking around the code base) was adjusting the `USERDATA_MAX_LENGTH` constant in `werift-webrtc/packages/sctp/src/sctp.ts`. It was set to `1200` and adjusting to `5500` gave me a good boost (about 3-4x improvement) in speed, but still quite far away from the performance of the `wrtc` node library (I can't wait to get rid of the wrtc library and its node-gyp dependencies (sigh))

My theory on why updating the constant worked is as follows. In my app I grab a file off the local computer and read with `fs` library. I break up the binary in to 5500 byte chunks and send to the datachannel. When the `USERDATA_MAX_LENGTH` was 1200, it was then breaking up the packets again before actually sending the file over the network.

Interestingly enough, if i adjusted the `USERDATA_MAX_LENGTH` constant too high the "performance boost" went away.

Thanks again for all the work here, truly impressive stuff!

**Comment by koush** [comment]
> `wrtc` node library (I can't wait to get rid of the wrtc library and its node-gyp dependencies (sigh))

I have a wrtc fork that contains prebuilts for every platform so you can avoid the full webrtc build process. There was some work to get it merged into the official repository but it seems abandoned.

https://github.com/koush/node-webrtc
https://github.com/node-webrtc/node-webrtc/pull/714

**Comment by cw-rterm** [comment]
Oh wow, very nice. The two platforms that I would really like are alpine linux and Apple Silicon/Arm (M1/M2, etc). wrtc worked with Rosetta, but that doesn't quite feel right.

I should also note that even wrtc datachannel speeds tap out around 3Mbps (at least for me). Pion and others (some ORTC libraries) report much higher numbers. I attempted to get Pion working with WASM, but that didn't quite go as planned. I didn't put much behind it; after I found this library it felt like a better fit for my specific project/use case.

Yes, I do hope to move to this package, it solved a lot of issues actually. For example, with wrtc the peer that opens the data channel doesn't get the `open` event to fire for the datachannel. There are other bugs that I had to work around over the last few years/months.

**Comment by cw-rterm** [comment]
> Fixed the nack handler to use a resetting timeout when a nack actually needs to happen.
> 
> <img alt="image" width="421" src="https://user-images.githubusercontent.com/73924/212202441-e8f4645d-c7b2-4fc3-8348-708984eb7e2b.png">

Is there any way to try this out or get a code sample to adjust things locally?

**Comment by tsightler** [comment]
I had hacked around a little with this wrtc build:
https://www.npmjs.com/package/@cubicleai/wrtc

But I don't really want to do all the work to move because I've also found it quite buggy and it seems mostly abandoned, also seeing that werift mostly works now, just not with ideal performance, it's hard to be motivated to do the work to implement something else entirely.

**Comment by tsightler** [comment]
> Taking a look at the js profiler, the werift nack loop is unnecessarily running 200 times per second, even when there is no packet loss. The top `anonymous` hit:

Interesting find, in the code that I have installed (0.17.5) it appears it's only 50 times per second (not that this is good), and that appears to be a very recent change.  Here's what it looks like 0.17.5:

```
  private start() {
    if (this.nackLoop) {
      return;
    }
    this.nackLoop = setInterval(async () => {
      try {
        await this.sendNack();
      } catch (error) {
        log("failed to send nack", error);
      }
    }, 20);
  }

```
And it was commit https://github.com/shinyoshiaki/werift-webrtc/commit/5aded34516535e0afa8692daac1a0346b07799ed where it was change to 5 just two weeks ago.  Clearly still should be fixed, but I guess that's why I had not noticed this on any profiles I had run.

**Comment by cw-rterm** [comment]
Oh nice! Where is 0.17.5? I only see 0.17.0 in releases/tags. Do i need to look through some commit history and the package.json files?

<img width="560" alt="image" src="https://user-images.githubusercontent.com/99921455/212345970-176fbf33-4c08-4fe7-a629-2a13433f5489.png">

**Comment by koush** [comment]
I'm working off the primary forked development branch: the 5ms is there:
https://github.com/shinyoshiaki/werift-webrtc/blob/b0f973191f8d9905d619f4488087f75215f76385/packages/webrtc/src/media/receiver/nack.ts#L38

Not sure what's in the published release.

Here's what's in my perf improvements branch thus far. https://github.com/koush/werift-webrtc/commit/c80d48773ccff60c5f105d292719f4c0c05073d0

**Comment by tsightler** [comment]
Thanks very much @koush, I really appreciate the work to optimize werift as much as reasonably possible.

**Comment by koush** [comment]
The other issue I've found confirms that werift is spending a lot of time in the actual encryption call, which are synchronous calls that could be asynchronous and multithreaded (but node does not provide an API for this). aes-gcm in SRTP lends itself well to mutithreading since it's not stream based. Keyframes can be 300k or more at times, which is 200 encrypt invocations. While moving this into a worker pool will smooth things out on main thread contention, it won't improve the actual CPU time used.
I suspect that the js marshaling to node native code and openssl 200 times uses a good chunk of CPU time. Doing this as a bulk transaction would likely improve things but is not possible without a custom native dependency. I'm also uncertain if node's openssl has hardware acceleration for OpenSSL EVP.

**Comment by tsightler** [comment]
Well, at least I wasn't totally off base!  I had thought of the idea of pre-allocating decrypt objects asynchronously in another worker thread since the counter used for the IV is predictable.  I thought that might at least improve on the current in-line allocation, but I questioned if it would really help since, as you noted, it wouldn't really lower the amount of total processing.

**Comment by koush** [comment]
> pre-allocating decrypt objects asynchronously in another worker thread since the counter used for the IV is predictable

I believe OpenSSL EV functions allow resetting the IV (ie, reusing the crypto session), but that's not in the node api unfortunately.

**Comment by cw-rterm** [comment]
Been messing around (what I think is) the send/receive code today (`/packages/sctp/src/sctp.ts`), mainly logging to see what (if anything) I might find. The one thing I did notice is the queue does a "burst" and waits until the end before sending the next frame of packets. Is this a requirement or expected or is there a way to gently add to the queue as packets are sent (not waiting for an entire batch to send).

**Comment by koush** [comment]
I'm not familiar with scp implementation details but there definitely seems to be a 3 second sleep in there. https://github.com/shinyoshiaki/werift-webrtc/blob/b0f973191f8d9905d619f4488087f75215f76385/packages/sctp/src/sctp.ts#L987

**Comment by Zubnix** [comment]
I'm still experiencing quite horrible datachannel performance with the currently released version. Looking at the implementation I  noticed an excessive use of of async/await stemming from the use of (an effectively needless) async send in the udp ice transport. Investigating further I found that the timer3Handle (and related usage) for some reason makes the whole send stream stutter heavily.

As a test I removed the async/awaits from the sctp implementation, and completely removed the timer3handle. In my own project and subjective testing I noticed that sending is now faster than node-webrtc or node-datachannel. Probably because removing this removes sctp's absolutely horrible congestion control & resend mechanism I guess (which I don't need for my use-case anyway).

Diff: https://github.com/shinyoshiaki/werift-webrtc/compare/develop...Zubnix:werift-webrtc:better-datachannel-perf in case people are interested.

**Comment by cw-rterm** [comment]
Amazing! Can't wait to try out the update

**Comment by cw-rterm** [comment]
I tried to do this today, but ran in to all kinds of problems with installing the module. It seems that even `npm install werift` is broken for me.

<img width="480" alt="image" src="https://user-images.githubusercontent.com/99921455/222802148-faa49a87-3f6e-4656-a719-9b87b62f7149.png">


To try the patch i tried `npm i git+https://github.com/shinyoshiaki/werift-webrtc.git#c18af7100212c8b94ee469b780e4328f33fdb438`

Likely something simple on my end, but do i need to do anything in the `node_modules/werift` folder after this? `yarn` and/or `yarn build`.

trying to run `yarn build` has its own errors

<img width="1270" alt="image" src="https://user-images.githubusercontent.com/99921455/222803031-48ba074b-a3e9-446c-8129-5f8b9bfcf6ce.png">

anyway, look forward to trying it out once i can get these sorted. If any one has tips, i would appreciate it!

**Comment by koush** [comment]
Looking into this again.

The data channel throughput seems to be limited by he message size. If I use messages of size 8 the bitrate barely exceeds 2000 bps. Extremely slow. If I use a message of size 32k, the bitrate gets up to around 3Mbps max. I am unable to exceed 3Mbps even locally. 

After some more investigation, the transport seems to end up in a degraded state and this line takes up to 9ms at times:
https://github.com/shinyoshiaki/werift-webrtc/blob/89af3f1e259b3bcfb4b08da8ee5e085a91640de8/packages/webrtc/src/transport/sctp.ts#L284

Node/v8 may be doing a full array copy and gc or something under the hood.

Using a for (const chunk of this.outboundQueue) speeds this up significantly.

**Comment by koush** [comment]
I am able to get 5Mbit max bidirectional (M1 Studio Ultra), or 10Mbit one way. This code is CPU bound likely due to dtls being synchronous (one of my cores stays at 100%). 10Mbit throughput is roughly 1000 packets per second. There needs to be a way to get this off the main thread.

Because dtls is connectionless, encrypting and sending packets out of different threads with no concern for order (let the other side figure it out) is a viable strategy.

That would double the overall send throughput.

I'll see if I can prototype a thread pool approach.

**Comment by koush** [comment]
I've got a prototype that switches the encrypt call from node crypto to web crypto (which is promise based and asynchronous), this is around 30% faster, reaching 13Mbit. The ramp up time is also significantly faster.

Using web crypto for decrypt, however, significantly slows it down, around 50%. I suspect because decrypt is not CPU intensive so the promise based approach is not worthwhile.

**Comment by koush** [comment]
Found a major issue. This change got a 6x improvement. Throughput is around 38Mbit now (bidirectional) where it was 5.5Mbit prior..

[1660dd5d](https://github.com/shinyoshiaki/werift-webrtc/pull/454)
